core/tcp_connection.cpp

```diff
@@ -1,36 +1,279 @@
 #include "tcp_connection.h"
 #include "global_include.h"
 
+#ifndef WINDOWS
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <errno.h>
+#include <unistd.h> // for close()
+#else
+#include <winsock2.h>
+#include <Ws2tcpip.h> // For InetPton
+#undef SOCKET_ERROR // conflicts with ConnectionResult::SOCKET_ERROR
+#endif
+
+#include <cassert>
+
+#ifndef WINDOWS
+#define GET_ERROR(_x) strerror(_x)
+#else
+#define GET_ERROR(_x) WSAGetLastError()
+#endif
+
 namespace dronecore {
 
 TcpConnection::TcpConnection(DroneCoreImpl *parent, const std::string &ip, int port_number) :
-    Connection(parent)
+    Connection(parent),
+    _remote_ip(ip),
+    _remote_port_number(port_number),
+    _should_exit(false)
+{
+    if (_local_port_number == 0) {
+        _local_port_number = DEFAULT_TCP_LOCAL_PORT;
+    }
+}
+
+TcpConnection::~TcpConnection()
 {
-    UNUSED(ip);
-    UNUSED(port_number);
+    // If no one explicitly called stop before, we should at least do it.
+    stop();
 }
 
+
 bool TcpConnection::is_ok() const
 {
     return true;
 }
 
 DroneCore::ConnectionResult TcpConnection::start()
 {
+    if (!start_mavlink_receiver()) {
+        return DroneCore::ConnectionResult::CONNECTIONS_EXHAUSTED;
+    }
+
+    DroneCore::ConnectionResult ret = setup_port();
+    if (ret != DroneCore::ConnectionResult::SUCCESS) {
+        return ret;
+    }
+
+    start_recv_thread();
+
     return DroneCore::ConnectionResult::SUCCESS;
 }
 
+DroneCore::ConnectionResult TcpConnection::setup_port()
+{
+
+#ifdef WINDOWS
+    WSADATA wsa;
+    if (WSAStartup(MAKEWORD(2, 2), &wsa) != 0) {
+        Debug() << "Error: Winsock failed, error: %d", WSAGetLastError();
+        return DroneCore::ConnectionResult::SOCKET_ERROR;
+    }
+#endif
+
+    _socket_fd = socket(AF_INET, SOCK_STREAM, 0);
+
+    if (_socket_fd < 0) {
+        Debug() << "socket error" << GET_ERROR(errno);
+        return DroneCore::ConnectionResult::SOCKET_ERROR;
+    }
+
+    sockaddr_in local_addr;
```

Comment:
In C++ you can just initialize to 0 without memset doing: `sockaddr_in local_addr {};`